Add ability for a user to disable their Gravatar.

stevepolitodesign opened the pull request:
This is in reference to #82. Specifically, it allows a user to decide whether or not their Gravatar (or placeholder image) appears in the navigation. This is set in the registration form, and is saved in a `show_gravatar` column on the `users` table.

excid3 commented:
Thanks @stevepolitodesign 👍 

I think the direction I wanted to go with this was more along the lines of:

1. Use gravatar as the fallback
2. Allow users to upload avatars with ActiveStorage
3. Allow users to change their ActiveStorage avatar
4. Allow users to remove their ActiveStorage avatar (which should then show their gravatar avatar as the fallback).

We would probably just want an avatar helper that looks at the user, checks if the have an ActiveStorage uploader and uses that if it exists, otherwise falls back to rendering the Gravatar image url if it doesn't exist.

Does that make sense?

stevepolitodesign commented:
@excid3 thanks for the feedback! I'll continue to work on this. Since I'm new to PR's should we close this PR? Or, should I revert my changes, and then continue to build out the new features here?

excid3 commented:
If you want, keep building them in this branch. You'll want to remove these
changes though. Whichever is easier for you. 👍

On Sat, May 16, 2020 at 7:18 AM Steve Polito Design <
notifications@github.com> wrote:

> @excid3 <https://github.com/excid3> thanks for the feedback! I'll
> continue to work on this. Since I'm new to PR's should we close this PR?
> Or, should I revert my changes, and then continue to build out the new
> features here?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/excid3/jumpstart/pull/105#issuecomment-629636870>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAQMFICPFLQIJIH2M3DNE3RRZ775ANCNFSM4NCRFMOA>
> .
>
-- 
Chris Oliver
http://excid3.com | @excid3 <http://twitter.com/excid3> |
http://github.com/excid3

stevepolitodesign commented:
> We would probably just want an avatar helper that looks at the user, checks if the have an ActiveStorage uploader and uses that if it exists, otherwise falls back to rendering the Gravatar image url if it doesn't exist.

Should I be handling the [ActiveStorage setup](https://guides.rubyonrails.org/active_storage_overview.html#setup) and as well as [attaching the avatar to the user](https://guides.rubyonrails.org/active_storage_overview.html#attaching-files-to-records) in the `template.rb` file?

 Or, is it as simple as conditionally adding a file field to the registration templates if the `user` model has a `has_one_attached` method?

excid3 commented:
@stevepolitodesign Good questions! I think that we probably do need to:

1. Run the activestorage install
2. Add has_one_attached :avatar to User (you can use the `rails g model User avatar:file` to have this added automatically... I actually added that feature to Rails!)
3. Add the avatar file field to the edit user registration template so users can change it. Also nice if the user can delete their avatar.
4. Add a helper for rendering the avatar that:
  * Checks for attached avatar, uses that. Falls back to Gravatar if not.

And we can use the Ruby MD5 hash to generate the gravatar URL without having to use a gem for it. 👍 

Is that enough to get started in the right direction? Happy to add any more details you might need!

stevepolitodesign commented:
@excid3 thanks for the detailed response! This looks like enough for me to get started, but I'll ping you if I have any questions. Maybe I'll extract this into an Avatar generator on RailsBytes when it's all said and done.

excid3 commented:
An avatar generator is a great idea!

stevepolitodesign commented:
@excid3 I went ahead and made the following changes:

- Update `app/models/user.rb` to include `has_one_attached :avatar`.
- Add `rails_command "active_storage:install"` to `template.rb`.
- Remove `gravatar_image_tag` gem and corresponding `config/initializers/gravatar.rb`.
- Create helper `avatar_path` helper method that returns a url to the object's avatar, with a fallback to their Gravatar.
- Require `bs-custom-file-input` as a dependency, and load into `app/javascript/packs/application.js`.
- Add an avatar field to `app/views/devise/registrations/edit.html.erb`.
- Add a preview of the users avatar to `app/views/devise/registrations/edit.html.erb`.

excid3 commented:
@stevepolitodesign It's been a month already? Dang! I'm digging into this now. 🤘

Hope your new job has been going well!

excid3 commented on app/helpers/application_helper.rb:
```diff
@@ -7,4 +7,20 @@ def bootstrap_class_for(flash_type)
+      gravatar_id = Digest::MD5::hexdigest(object.email.downcase)
+      "#{base_url}/#{gravatar_id}#{base_url_params}"
+    else
+      "#{base_url}/00000000000000000000000000000000#{base_url_params}"
```
Just curious, what's this do? Is this the default image?

stevepolitodesign replied:
Hi Chris! Thanks for taking the time to review this PR. Yes, `"#{base_url}/00000000000000000000000000000000#{base_url_params}"` returns the [default image](https://en.gravatar.com/site/implement/images/).

excid3 commented on app/javascript/packs/application.js:
```diff
@@ -13,8 +13,10 @@ window.Rails = Rails
 $(document).on("turbolinks:load", () => {
   $('[data-toggle="tooltip"]').tooltip()
   $('[data-toggle="popover"]').popover()
+  bsCustomFileInput.init()
```
Do you think we need this library? I'd prefer to maybe not include extra dependencies and I don't think we need a fancy file input field for the base Jumpstart template. Maybe we could add this to the Wiki instead?

stevepolitodesign replied:
Sounds good, I can make this change this week 👍

stevepolitodesign commented:
@excid3 I went ahead and removed the dependency, and adjusted the file field styles accordingly. I recorded my screen to show you how this new feature will work on [sign up](https://share.getcloudapp.com/wbuW6b4Z). I also recorded my screen to demonstrate that [the default Gravatar](https://share.getcloudapp.com/2Nu5ql44) will load if a user does not have a Gravatar.

excid3 commented:
@stevepolitodesign awesome, this looks fantastic. Gonna merge it 🤘  Thanks for all this!
